### ISSUE TITLE
Mega Man 3 Improvement: Wobbling an Select Screen

### ISSUE THREAD

**Issue by xadox-1st** [SourMesen/Mesen#630]
The Rom-Hack [Mega Man 3 Improvement](http://www.romhacking.net/hacks/992/) shows a wobble in the lower part at Stage Select Screen and at weapon select screen.

[Megaman III (U) [Improvement Ver 2.10]-190528-112238.zip](https://github.com/SourMesen/Mesen/files/3226662/Megaman.III.U.Improvement.Ver.2.10.-190528-112238.zip) 

[Megaman III (U) [Improvement Ver 2.10]-190528-111715.zip](https://github.com/SourMesen/Mesen/files/3226633/Megaman.III.U.Improvement.Ver.2.10.-190528-111715.zip)

**Comment by retro-wertz** [comment]
i think this is generally a timing problem with the mapper. using verified
rom dump of Megame a line is visible in select screen on top of shadow man

**Comment by SourMesen** [comment]
This hack has already been mentioned to me in the past, but unless someone can actually test it on hardware (e.g: a north america NES with a real original MMC3 board), there isn't much to be done - it's very possible the hack itself has incorrect timings that were tested against emulators only.

**Comment by RokkumanX** [comment]
The so called "wobbling" is present on real hardware:
https://www.youtube.com/watch?v=RAmnTuCUqpg

I know the ones behind the video "Japanspel", translated "Japan games" because I frequently buy Japanese games from them, including my Japanese Famicom/Famicom Disk System collection.

Even when I was a kid owning the PAL version the "wobbling" was present. Every accurate emulator emulates this correctly.

My dumped copy of Rockman 3 is emulated exactly as showed in the above video.

**Comment by xadox-1st** [comment]
Yes like @retro-wertz already mentioned the one line above _Shadow Man_ occurs on an unpached rom, or original hardware, also.

Version 1.3 of the patch fixes this problem also on real hardware.
That's why I've always used the patch in the past.

I now have to test if the current version 2.10 of the patch causes the wobble problem on real hardware, also.

**Comment by RokkumanX** [comment]
Oh.....we were talking about a Rom-Hack, well that's way out of my jurisdiction.

**Comment by xadox-1st** [comment]
> This hack has already been mentioned to me in the past, but unless someone can actually test it on hardware (e.g: a north america NES with a real original MMC3 board), there isn't much to be done - it's very possible the hack itself has incorrect timings that were tested against emulators only.

I tried the Rom Hack yesterday on an NTSC front loader with an Everdrive N8.
The problem with the wiggle doesn't occur here.

**Comment by domgetter** [comment]
> I tried the Rom Hack yesterday on an NTSC front loader with an Everdrive N8.
> The problem with the wiggle doesn't occur here.

Keep in mind that Rom loaders like Everdrive and Powerpak have to re-implement MMC3 and the IRQ timing is surprisingly difficult to get right (often the IRQ will be a couple CPU cycles early or late).  The original boards often slapped a capacitor on the board to smooth out the noisy PPU address lines.

See https://forums.nesdev.com/viewtopic.php?f=9&t=11945&start=29 (emphasis mine)

> New MMC3 IRQ implementation (RAMBO-1 style, based on M2). **Doesn't
  exactly match real MMC3 timings**, but should be more robust than the
  previous implementation.

For example, Startropics often have sprite flicker for the main character on Everdrive and Powerpack but not on the original cartridges.

The "correct" test would be to use a board with a genuine MMC3 chip (as SourMesen stated above), and see if the Romhack has the same behavior.

**Comment by CharlesCorrin** [comment]
One of the changes made on November 11th has fixed this bug. Congratulations on the continued great work. :3

**Comment by CharlesCorrin** [comment]
Aha, something else. The hack will also randomly bug out big when you press start during a stage to open your weapons menu. It flashes numbers and letters on the lower portion of the screen. It doesn't always do this, but it does sometimes. Just as the hack randomly doesn't do the jiggle thing on the stage select.

**Comment by CharlesCorrin** [comment]
A contact of mine messaged the writer of this hack. But I've been too busy to post again till now. Sorry.

Anyway, the hack's writer assured that the problem is with Mesen and not with the game hack. They say the real NES and other emulators (Nestopia, Nintendulator, Fceux) don't have the bug, but also that he wrote it to match real hardware rather than emulators. That the bug has to do with IRQ inaccuracies. Unfortunately, they couldn't run Mesen to debug because it won't run on their old OS.

Perhaps it's a topic for NesDev to analyze?

**Comment by domgetter** [comment]
>They say the real NES and other emulators (Nestopia, Nintendulator, Fceux) don't have the bug,

Was this tested on a real MMC3 board, or just a Powerpack/Everdrive?

**Comment by SourMesen** [comment]
RE: The behavior you've described before, I'm assuming you've enabled all the "not recommended" options in the emulation->advanced->homebrew bit? If so, that's is what is making the jitter effect random on the stage select random, and the glitch in the weapon menu is also caused by one of these options.

Both of these glitches are essentially caused by the code running ever-so-slightly too early or late, and the exact timing can change based on the random "alignment" you get on power on (which is one of the advanced options

Like domgetter above said, though, unless the hack was tested on an actual MMC3 chip (not a reproduction cart, not a flash cart), it's hard to know which behavior is correct (and it's possible the hardware sometimes works and sometimes jitters from one reset to another, too)

**Comment by CharlesCorrin** [comment]
I tested with the "not recommended" options AND without them. The wobbling happens on both. It also randomly doesn't happen with the options checked AND unchecked both. So there's no difference in the behavior with those options checked versus unchecked.

I suppose I'll open an account on NesDev later and we'll try to figure out what's wrong with Mesen or the hack there.
